core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -857,6 +858,364 @@ KOKKOS_INLINE_FUNCTION_DELETED Impl::ThreadVectorRangeBoundariesStruct<
+  static constexpr Iterate iter         = Iter;
+
+  static constexpr Iterate direction =
+      iter == Iterate::Default
+          ? layout_iterate_type_selector<array_layout>::outer_iteration_pattern
+          : iter;
+
+  template <class... Args>
+  KOKKOS_INLINE_FUNCTION MDThreadVectorRange(team_handle_type const& team_,
+                                             Args&&... args)
+      : team(team_), boundaries{static_cast<boundary_type>(args)...} {
+    static_assert(sizeof...(Args) == TotalRank);
+  }
+
+  team_handle_type const& team;
+  boundary_type boundaries[total_rank];
+};
+
+template <typename TeamHandle, typename... Args>
+MDThreadVectorRange(TeamHandle const&, Args&&...)
+    ->MDThreadVectorRange<Kokkos::Iterate::Default, sizeof...(Args),
+                          TeamHandle>;
+
+template <Iterate Iter, int TotalRank, typename TeamHandle>
+struct MDTeamVectorRange {
+  using rank_type        = int;
+  using boundary_type    = int;
+  using team_handle_type = TeamHandle;
+  using execution_space  = typename team_handle_type::execution_space;
+  using array_layout     = typename execution_space::array_layout;
+  static constexpr rank_type total_rank = TotalRank;
+  static constexpr Iterate iter         = Iter;
+
+  static constexpr Iterate direction =
+      iter == Iterate::Default
+          ? layout_iterate_type_selector<array_layout>::outer_iteration_pattern
+          : iter;
+
+  template <class... Args>
+  KOKKOS_INLINE_FUNCTION MDTeamVectorRange(team_handle_type const& team_,
+                                           Args&&... args)
+      : team(team_), boundaries{static_cast<boundary_type>(args)...} {
+    static_assert(sizeof...(Args) == TotalRank);
+  }
+
+  team_handle_type const& team;
+  boundary_type boundaries[total_rank];
+};
+
+template <typename TeamHandle, typename... Args>
+MDTeamVectorRange(TeamHandle const&, Args&&...)
+    ->MDTeamVectorRange<Kokkos::Iterate::Default, sizeof...(Args), TeamHandle>;
+
+template <Iterate Iter, int TotalRank, typename TeamHandle, typename Lambda,
+          typename ReducerValueType>
+KOKKOS_INLINE_FUNCTION void parallel_reduce(
+    MDTeamThreadRange<Iter, TotalRank, TeamHandle> const& policy,
+    Lambda const& lambda, ReducerValueType& val) {
+  Impl::MDTeamRangeStatus<Impl::MDTeamRangeLastRank::NotLastRank,
+                          Impl::MDTeamRangeParThread::ParThread,
+                          Impl::MDTeamRangeParVector::NotParVector>
+      rangeStatus;
+
+  Impl::md_parallel_impl<Impl::MDTeamRangeThreadAndVector::NotBoth>(
+      policy, lambda, rangeStatus, val);
+}
+
+// FIXME Check if passing in NoReductionTag by value vs reference shows any
+// difference
+template <Iterate Iter, int TotalRank, typename TeamHandle, typename Lambda>
+KOKKOS_INLINE_FUNCTION void parallel_for(
+    MDTeamThreadRange<Iter, TotalRank, TeamHandle> const& policy,
+    Lambda const& lambda) {
+  parallel_reduce(policy, lambda, Impl::no_reduction_tag);
```

Comment:
Something in me doesn't like this and would prefer calling the impl_parallel function directly, but I guess its a question of code duplication?

Reply:
Also changed in [c534937](https://github.com/kokkos/kokkos/pull/5238/commits/c5349378887bbbf53a29cd76eb21095cdd18b83d)